buildSrc/src/main/kotlin/dev/zacsweers/catchup/gradle/LicensesJsonGenerator.kt

```diff
@@ -0,0 +1,142 @@
+package dev.zacsweers.catchup.gradle
+
+import com.android.build.gradle.AppExtension
+import com.squareup.moshi.JsonWriter
+import okio.buffer
+import okio.sink
+import org.gradle.api.DefaultTask
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.artifacts.ArtifactView
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier
+import org.gradle.api.artifacts.result.ResolvedArtifactResult
+import org.gradle.api.attributes.Attribute
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.Classpath
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+import org.gradle.kotlin.dsl.findByType
+import org.gradle.kotlin.dsl.register
+import org.gradle.maven.MavenModule
+import org.gradle.maven.MavenPomArtifact
+import org.w3c.dom.Node
+import org.w3c.dom.NodeList
+import java.util.Locale
+import javax.inject.Inject
+import javax.xml.parsers.DocumentBuilderFactory
+
+class LicensesJsonGeneratorPlugin : Plugin<Project> {
+  override fun apply(project: Project) {
+    project.extensions.findByType<AppExtension>()!!.applicationVariants.configureEach {
+      if (buildType.name == "release") {
+        project.tasks.register<LicensesJsonGenerator>("generateLicensesJson") {
+          group = "licenses"
+          artifactFiles.from(runtimeConfiguration.artifactView().artifacts.artifactFiles)
+          configuration = runtimeConfiguration
+          jsonFile.set(project.file("licenses.json"))
+        }
+      }
+    }
+  }
+
+  private fun Configuration.artifactView(): ArtifactView {
+    return incoming.artifactView {
+      attributes {
+        attribute(Attribute.of("artifactType", String::class.java), "android-classes")
+      }
+    }
+  }
+}
+
+open class LicensesJsonGenerator @Inject constructor(objects: ObjectFactory) : DefaultTask() {
```

Comment:
I'm not sure where the docs are on this (but they do exist). You can instead declare this class as
```kotlin
abstract class LicensesJsonGenerator : DefaultTask()
```
salient points:
1. it's abstract instead of open
2. It doesn't define any constructor

With that, you can then do this for your properties:
```kotlin
@get:Classpath
abstract val artifactFiles: ConfigurableFileCollection
```
salient points:
1. the property is abstract
2. you no longer need to initialize it manually from ObjectFactory

This works because Gradle extends all this stuff at runtime. E.g., it'll create a class named `LicensesJsonGenerator_Decorated`. It uses that decorated version.

Basically, Gradle already "knows" how to initialize a `ConfigurableFileCollection`. It can do automatically what you've done manually. Benefit is a bit less boilerplate when defining your task classes.

Please note that this is fine and you should leave it as-is:
```kotlin
  @get:Internal
  lateinit var configuration: Configuration
```

Reply:
good shout, done here f2253a7